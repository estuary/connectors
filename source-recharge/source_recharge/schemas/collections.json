{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "required": ["id"],
  "properties": {
    "id": {
      "type": ["integer"]
    },
    "name": {
      "type": ["null", "string"]
    },
    "created_at": {
      "type": ["null", "string"],
      "format": "date-time"
    },
    "updated_at": {
      "type": ["null", "string"],
      "format": "date-time"
    }
  }
}
