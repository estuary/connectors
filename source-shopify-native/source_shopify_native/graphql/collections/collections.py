from datetime import datetime
from logging import Logger
import json
from typing import Any, AsyncGenerator, ClassVar, Literal

from ...models import ShopifyGraphQLResource, SortKey


class _Collections(ShopifyGraphQLResource):
    COLLECTION_TYPE: ClassVar[str] = ""
    QUERY_ROOT = "collections"
    SORT_KEY = SortKey.UPDATED_AT
    QUERY = """
    title
    handle
    description
    descriptionHtml
    templateSuffix
    sortOrder
    productsCount {
        count
        precision
    }
    products {
        edges {
            node {
                id
                legacyResourceId
            }
        }
    }
    publications {
        edges {
            node {
                __typename
                publishDate
                publication {
                    id
                    name
                }
            }
        }
    }
    """

    def __init_subclass__(cls, **kwargs: Any) -> None:
        super().__init_subclass__(**kwargs)
        if not cls.COLLECTION_TYPE:
            raise NotImplementedError("Subclasses must set COLLECTION_TYPE")

    @staticmethod
    def build_query_with_type(
        collection_type: Literal["custom", "smart"], start: datetime, end: datetime
    ) -> str:
        return _Collections.build_query_with_fragment(
            start,
            end,
            query=f"collection_type:{collection_type}",
            includeCreatedAt=False,
        )

    @staticmethod
    def build_query(start: datetime, end: datetime) -> str:
        raise NotImplementedError("build_query method must be implemented")

    @staticmethod
    async def process_result(
        log: Logger, lines: AsyncGenerator[bytes, None]
    ) -> AsyncGenerator[dict, None]:
        PRODUCTS_KEY = "products"
        PUBLICATIONS_KEY = "publications"
        current_collection = None

        async for line in lines:
            record: dict[str, Any] = json.loads(line)
            id: str = record.get("id", "")
            parent_id = record.get("__parentId", "")
            typename = record.get("__typename", "")

            if "gid://shopify/Collection/" in id:
                if current_collection:
                    yield current_collection

                current_collection = record
                current_collection[PRODUCTS_KEY] = []
                current_collection[PUBLICATIONS_KEY] = []

            elif "gid://shopify/Product/" in id:
                if not current_collection:
                    log.error("Found a product before finding a collection.")
                    raise RuntimeError()
                if parent_id != current_collection.get("id", ""):
                    log.error(
                        "Product's parent id does not match the current collection's id.",
                        {
                            "product.id": id,
                            "product.__parentId": parent_id,
                            "current_collection.id": current_collection.get("id"),
                        },
                    )
                    raise RuntimeError()

                current_collection[PRODUCTS_KEY].append(record)

            elif typename == "CollectionPublication":
                if not current_collection:
                    log.error("Found a publication before finding a collection.")
                    raise RuntimeError()
                if parent_id != current_collection.get("id", ""):
                    log.error(
                        "Publication's parent id does not match the current collection's id.",
                        {
                            "publication.__parentId": parent_id,
                            "current_collection.id": current_collection.get("id"),
                        },
                    )
                    raise RuntimeError()

                record.pop("__typename", None)
                current_collection[PUBLICATIONS_KEY].append(record)

            else:
                log.error("Unidentified line in JSONL response.", record)
                raise RuntimeError()

        if current_collection:
            yield current_collection


class CustomCollections(_Collections):
    NAME = "custom_collections"
    COLLECTION_TYPE = "custom"

    @staticmethod
    def build_query(start: datetime, end: datetime) -> str:
        return _Collections.build_query_with_type("custom", start, end)

    @staticmethod
    def process_result(
        log: Logger, lines: AsyncGenerator[bytes, None]
    ) -> AsyncGenerator[dict, None]:
        return _Collections.process_result(log, lines)


class SmartCollections(_Collections):
    NAME = "smart_collections"
    COLLECTION_TYPE = "smart"

    @staticmethod
    def build_query(start: datetime, end: datetime) -> str:
        return _Collections.build_query_with_type("smart", start, end)

    @staticmethod
    def process_result(
        log: Logger, lines: AsyncGenerator[bytes, None]
    ) -> AsyncGenerator[dict, None]:
        return _Collections.process_result(log, lines)
