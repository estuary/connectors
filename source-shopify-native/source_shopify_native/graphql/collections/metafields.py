from datetime import datetime
from logging import Logger
from typing import AsyncGenerator
from ..metafields import MetafieldsResource

from source_shopify_native.models import SortKey


class CustomCollectionMetafields(MetafieldsResource):
    NAME = "custom_collection_metafields"
    PARENT_ID_KEY = "gid://shopify/Collection/"
    QUERY_ROOT = "collections"
    SORT_KEY = SortKey.UPDATED_AT

    @staticmethod
    def build_query(
        start: datetime,
        end: datetime,
        first: int | None = None,
        after: str | None = None,
    ) -> str:
        return CustomCollectionMetafields.build_query_with_fragment(
            start,
            end,
            query="AND collection_type:custom",
            first=first,
            after=after,
            includeCreatedAt=False,
        )

    @staticmethod
    def process_result(
        log: Logger, lines: AsyncGenerator[bytes, None]
    ) -> AsyncGenerator[dict, None]:
        return MetafieldsResource._process_metafields_result(
            log, lines, CustomCollectionMetafields.PARENT_ID_KEY
        )


class SmartCollectionMetafields(MetafieldsResource):
    NAME = "smart_collection_metafields"
    PARENT_ID_KEY = "gid://shopify/Collection/"
    QUERY_ROOT = "collections"
    SORT_KEY = SortKey.UPDATED_AT

    @staticmethod
    def build_query(
        start: datetime,
        end: datetime,
        first: int | None = None,
        after: str | None = None,
    ) -> str:
        return SmartCollectionMetafields.build_query_with_fragment(
            start,
            end,
            query="AND collection_type:smart",
            first=first,
            after=after,
            includeCreatedAt=False,
        )

    @staticmethod
    def process_result(
        log: Logger, lines: AsyncGenerator[bytes, None]
    ) -> AsyncGenerator[dict, None]:
        return MetafieldsResource._process_metafields_result(
            log, lines, SmartCollectionMetafields.PARENT_ID_KEY
        )
