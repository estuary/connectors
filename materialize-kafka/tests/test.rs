use core::str;
use std::collections::HashMap;

use rdkafka::{
    admin::{AdminClient, AdminOptions},
    consumer::{BaseConsumer, Consumer},
    ClientConfig, Message, TopicPartitionList,
};

const BOOTSTRAP_SERVERS: &str = "localhost:9092";
const SCHEMA_REGISTRY_ENDPOINT: &str = "http://localhost:8081";
static KAFKA_INTERNAL_TOPICS: [&str; 4] = [
    "__consumer_offsets",
    "__amazon_msk_canary",
    "_schemas",
    "__transaction_state",
];

#[test]
fn test_spec() {
    let output = std::process::Command::new("flowctl")
        .args([
            "raw",
            "spec",
            "--source",
            "tests/test.flow.yaml",
            "--name",
            "acmeCo/materialize-kafka/avro",
        ])
        .output()
        .unwrap();

    assert!(output.status.success());
    let got: serde_json::Value = serde_json::from_slice(&output.stdout).unwrap();
    insta::assert_snapshot!(serde_json::to_string_pretty(&got).unwrap());
}

#[tokio::test]
async fn test_materialization() {
    drop_topics().await;

    for name in [
        "acmeCo/materialize-kafka/avro",
        "acmeCo/materialize-kafka/json",
    ] {
        let output = std::process::Command::new("flowctl")
            .args([
                "preview",
                "--source",
                "tests/test.flow.yaml",
                "--fixture",
                "tests/fixture.json",
                "--name",
                name,
            ])
            .output()
            .unwrap();

        println!("{}", str::from_utf8(&output.stderr).unwrap());
        assert!(output.status.success());
    }

    insta::assert_snapshot!(snapshot_topics().await);
}

async fn drop_topics() {
    let admin: AdminClient<_> = ClientConfig::new()
        .set("bootstrap.servers", BOOTSTRAP_SERVERS)
        .create()
        .unwrap();

    let topics = list_topics().await;
    let topics = topics.iter().map(String::as_str).collect::<Vec<&str>>();
    if !topics.is_empty() {
        admin
            .delete_topics(&topics, &AdminOptions::default())
            .await
            .unwrap();
    }
}

async fn list_topics() -> Vec<String> {
    let consumer: BaseConsumer = ClientConfig::new()
        .set("bootstrap.servers", BOOTSTRAP_SERVERS)
        .create()
        .unwrap();

    let meta = consumer.fetch_metadata(None, None).unwrap();
    let mut topics = meta
        .topics()
        .iter()
        .filter(|t| !KAFKA_INTERNAL_TOPICS.contains(&t.name()))
        .map(|t| t.name().to_string())
        .collect::<Vec<_>>();
    topics.sort();
    topics
}

async fn snapshot_topics() -> String {
    let mut out = String::new();

    for topic in list_topics().await {
        if !out.is_empty() {
            out.push_str("\n");
        }
        out.push_str("************\n");
        out.push_str(format!("Topic: {}\n", topic).as_str());
        out.push_str("************\n");
        out.push_str(&snapshot_topic(&topic).await);
        out.push_str("\n");
    }

    out
}

async fn snapshot_topic(topic: &str) -> String {
    let consumer: BaseConsumer = ClientConfig::new()
        .set("bootstrap.servers", BOOTSTRAP_SERVERS)
        .set("enable.auto.commit", "false")
        .set("group.id", "materialize-kafka")
        .set("auto.offset.reset", "earliest")
        .set("enable.partition.eof", "true")
        .create()
        .unwrap();

    let meta = consumer.fetch_metadata(Some(topic), None).unwrap();
    assert!(meta.topics()[0].partitions().len() != 0);
    let assignment =
        meta.topics()[0]
            .partitions()
            .iter()
            .fold(TopicPartitionList::new(), |mut tpl, p| {
                tpl.add_partition(topic, p.id());
                tpl
            });
    consumer.assign(&assignment).unwrap();
    let mut n = consumer.assignment().unwrap().count();

    let mut schema_cache: HashMap<u32, apache_avro::Schema> = HashMap::new();

    let mut rows = Vec::new();
    for msg in consumer.iter() {
        match msg {
            Ok(msg) => {
                let key = parse_datum(msg.key().unwrap(), &mut schema_cache).await;
                let value = parse_datum(msg.payload().unwrap(), &mut schema_cache).await;
                let row = format!("key: {}, value: {}", key, value);
                rows.push(row);
            }
            _ => {
                n -= 1;
                if n == 0 {
                    break;
                }
            }
        }
    }
    rows.sort();
    rows.join("\n")
}

async fn parse_datum(datum: &[u8], schema_cache: &mut HashMap<u32, apache_avro::Schema>) -> String {
    match datum[0] {
        0 => {
            let schema_id = u32::from_be_bytes(datum[1..5].try_into().unwrap());

            if !schema_cache.contains_key(&schema_id) {
                schema_cache.insert(schema_id, fetch_schema(schema_id).await);
            }

            let schema = schema_cache.get(&schema_id).unwrap();
            let avro_value = apache_avro::from_avro_datum(schema, &mut &datum[5..], None).unwrap();

            source_kafka::pull::avro_to_json(avro_value, schema)
                .unwrap()
                .to_string()
        }
        _ => str::from_utf8(datum).unwrap().to_owned(),
    }
}

async fn fetch_schema(id: u32) -> apache_avro::Schema {
    let schema_string = reqwest::get(format!(
        "{SCHEMA_REGISTRY_ENDPOINT}/schemas/ids/{id}/schema"
    ))
    .await
    .unwrap()
    .error_for_status()
    .unwrap()
    .text()
    .await
    .unwrap();

    apache_avro::Schema::parse_str(&schema_string).unwrap()
}
