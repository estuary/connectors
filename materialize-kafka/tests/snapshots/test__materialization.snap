---
source: tests/test.rs
expression: snapshot_topics().await
---
************
Topic: test_topic_avro
************
key: {"_flow_key":{"p1":1,"p2":"one","p3":"a"}}, value: {"_flow_extra":{"slash~1key":{"json":"\"one\""},"slash~1root":{"json":"true"},"value/nested":{"json":"\"a\""},"value/slash~1nested":{"json":"1.0"}},"canary":"amputation's","flow_document":"{\"_meta\":{\"uuid\":\"7545a800-1dda-11b2-8000-071353030311\"},\"canary\":\"amputation's\",\"id\":1,\"slash/key\":\"one\",\"slash/root\":true,\"value\":{\"nested\":\"a\",\"slash/nested\":1.0}}","flow_published_at":"1970-01-01T01:00:00.000000000Z","id":1}
key: {"_flow_key":{"p1":2,"p2":"two","p3":"b"}}, value: {"_flow_extra":{"slash~1key":{"json":"\"two\""},"slash~1root":{"json":"false"},"value/nested":{"json":"\"b\""},"value/slash~1nested":{"json":"2.0"}},"canary":"armament's","flow_document":"{\"_meta\":{\"uuid\":\"75de3e80-1dda-11b2-8000-071353030311\"},\"canary\":\"armament's\",\"id\":2,\"slash/key\":\"two\",\"slash/root\":false,\"value\":{\"nested\":\"b\",\"slash/nested\":2.0}}","flow_published_at":"1970-01-01T01:00:01.000000000Z","id":2}
key: {"_flow_key":{"p1":3,"p2":"three","p3":"c"}}, value: {"_flow_extra":{"slash~1key":{"json":"\"three\""},"slash~1root":{"json":"true"},"value/nested":{"json":"\"c\""},"value/slash~1nested":{"json":"3.0"}},"canary":"splatters","flow_document":"{\"_meta\":{\"uuid\":\"7676d500-1dda-11b2-8000-071353030311\"},\"canary\":\"splatters\",\"id\":3,\"slash/key\":\"three\",\"slash/root\":true,\"value\":{\"nested\":\"c\",\"slash/nested\":3.0}}","flow_published_at":"1970-01-01T01:00:02.000000000Z","id":3}
key: {"_flow_key":{"p1":4,"p2":"four","p3":"d"}}, value: {"_flow_extra":{"slash~1key":{"json":"\"four\""},"slash~1root":{"json":"false"},"value/nested":{"json":"\"d\""},"value/slash~1nested":{"json":"4.0"}},"canary":"strengthen","flow_document":"{\"_meta\":{\"uuid\":\"d70a1000-1de2-11b2-8000-071353030311\"},\"canary\":\"strengthen\",\"id\":4,\"slash/key\":\"four\",\"slash/root\":false,\"value\":{\"nested\":\"d\",\"slash/nested\":4.0}}","flow_published_at":"1970-01-01T02:00:00.000000000Z","id":4}
key: {"_flow_key":{"p1":5,"p2":"five","p3":"e"}}, value: {"_flow_extra":{"slash~1key":{"json":"\"five\""},"slash~1root":{"json":"true"},"value/nested":{"json":"\"e\""},"value/slash~1nested":{"json":"5.0"}},"canary":"Kringle's","flow_document":"{\"_meta\":{\"uuid\":\"d7a2a680-1de2-11b2-8000-071353030311\"},\"canary\":\"Kringle's\",\"id\":5,\"slash/key\":\"five\",\"slash/root\":true,\"value\":{\"nested\":\"e\",\"slash/nested\":5.0}}","flow_published_at":"1970-01-01T02:00:01.000000000Z","id":5}
key: {"_flow_key":{"p1":6,"p2":"six","p3":"f"}}, value: {"_flow_extra":{"slash~1key":{"json":"\"six\""},"slash~1root":{"json":"false"},"value/nested":{"json":"\"f\""},"value/slash~1nested":{"json":"6.0"}},"canary":"grosbeak's","flow_document":"{\"_meta\":{\"uuid\":\"d83b3d00-1de2-11b2-8000-071353030311\"},\"canary\":\"grosbeak's\",\"id\":6,\"slash/key\":\"six\",\"slash/root\":false,\"value\":{\"nested\":\"f\",\"slash/nested\":6.0}}","flow_published_at":"1970-01-01T02:00:02.000000000Z","id":6}

************
Topic: test_topic_json_1
************
key: {"id":1,"slash~1key":"one","value/nested":"a"}, value: {"canary":"amputation's","flow_document":"{\"_meta\":{\"uuid\":\"7545a800-1dda-11b2-8000-071353030311\"},\"canary\":\"amputation's\",\"id\":1,\"slash/key\":\"one\",\"slash/root\":true,\"value\":{\"nested\":\"a\",\"slash/nested\":1.0}}","flow_published_at":"1970-01-01T01:00:00.000000000Z","id":1,"slash~1key":"one","slash~1root":true,"value/nested":"a","value/slash~1nested":1.0}
key: {"id":2,"slash~1key":"two","value/nested":"b"}, value: {"canary":"armament's","flow_document":"{\"_meta\":{\"uuid\":\"75de3e80-1dda-11b2-8000-071353030311\"},\"canary\":\"armament's\",\"id\":2,\"slash/key\":\"two\",\"slash/root\":false,\"value\":{\"nested\":\"b\",\"slash/nested\":2.0}}","flow_published_at":"1970-01-01T01:00:01.000000000Z","id":2,"slash~1key":"two","slash~1root":false,"value/nested":"b","value/slash~1nested":2.0}
key: {"id":3,"slash~1key":"three","value/nested":"c"}, value: {"canary":"splatters","flow_document":"{\"_meta\":{\"uuid\":\"7676d500-1dda-11b2-8000-071353030311\"},\"canary\":\"splatters\",\"id\":3,\"slash/key\":\"three\",\"slash/root\":true,\"value\":{\"nested\":\"c\",\"slash/nested\":3.0}}","flow_published_at":"1970-01-01T01:00:02.000000000Z","id":3,"slash~1key":"three","slash~1root":true,"value/nested":"c","value/slash~1nested":3.0}
key: {"id":4,"slash~1key":"four","value/nested":"d"}, value: {"canary":"strengthen","flow_document":"{\"_meta\":{\"uuid\":\"d70a1000-1de2-11b2-8000-071353030311\"},\"canary\":\"strengthen\",\"id\":4,\"slash/key\":\"four\",\"slash/root\":false,\"value\":{\"nested\":\"d\",\"slash/nested\":4.0}}","flow_published_at":"1970-01-01T02:00:00.000000000Z","id":4,"slash~1key":"four","slash~1root":false,"value/nested":"d","value/slash~1nested":4.0}
key: {"id":5,"slash~1key":"five","value/nested":"e"}, value: {"canary":"Kringle's","flow_document":"{\"_meta\":{\"uuid\":\"d7a2a680-1de2-11b2-8000-071353030311\"},\"canary\":\"Kringle's\",\"id\":5,\"slash/key\":\"five\",\"slash/root\":true,\"value\":{\"nested\":\"e\",\"slash/nested\":5.0}}","flow_published_at":"1970-01-01T02:00:01.000000000Z","id":5,"slash~1key":"five","slash~1root":true,"value/nested":"e","value/slash~1nested":5.0}
key: {"id":6,"slash~1key":"six","value/nested":"f"}, value: {"canary":"grosbeak's","flow_document":"{\"_meta\":{\"uuid\":\"d83b3d00-1de2-11b2-8000-071353030311\"},\"canary\":\"grosbeak's\",\"id\":6,\"slash/key\":\"six\",\"slash/root\":false,\"value\":{\"nested\":\"f\",\"slash/nested\":6.0}}","flow_published_at":"1970-01-01T02:00:02.000000000Z","id":6,"slash~1key":"six","slash~1root":false,"value/nested":"f","value/slash~1nested":6.0}

************
Topic: test_topic_json_2
************
key: {"id":1,"slash~1key":"one","value/nested":"a"}, value: {"canary":"amputation's","flow_document":"{\"_meta\":{\"uuid\":\"7545a800-1dda-11b2-8000-071353030311\"},\"canary\":\"amputation's\",\"id\":1,\"slash/key\":\"one\",\"slash/root\":true,\"value\":{\"nested\":\"a\",\"slash/nested\":1.0}}","flow_published_at":"1970-01-01T01:00:00.000000000Z","id":1,"slash~1key":"one","slash~1root":true,"value/nested":"a","value/slash~1nested":1.0}
key: {"id":2,"slash~1key":"two","value/nested":"b"}, value: {"canary":"armament's","flow_document":"{\"_meta\":{\"uuid\":\"75de3e80-1dda-11b2-8000-071353030311\"},\"canary\":\"armament's\",\"id\":2,\"slash/key\":\"two\",\"slash/root\":false,\"value\":{\"nested\":\"b\",\"slash/nested\":2.0}}","flow_published_at":"1970-01-01T01:00:01.000000000Z","id":2,"slash~1key":"two","slash~1root":false,"value/nested":"b","value/slash~1nested":2.0}
key: {"id":3,"slash~1key":"three","value/nested":"c"}, value: {"canary":"splatters","flow_document":"{\"_meta\":{\"uuid\":\"7676d500-1dda-11b2-8000-071353030311\"},\"canary\":\"splatters\",\"id\":3,\"slash/key\":\"three\",\"slash/root\":true,\"value\":{\"nested\":\"c\",\"slash/nested\":3.0}}","flow_published_at":"1970-01-01T01:00:02.000000000Z","id":3,"slash~1key":"three","slash~1root":true,"value/nested":"c","value/slash~1nested":3.0}
key: {"id":4,"slash~1key":"four","value/nested":"d"}, value: {"canary":"strengthen","flow_document":"{\"_meta\":{\"uuid\":\"d70a1000-1de2-11b2-8000-071353030311\"},\"canary\":\"strengthen\",\"id\":4,\"slash/key\":\"four\",\"slash/root\":false,\"value\":{\"nested\":\"d\",\"slash/nested\":4.0}}","flow_published_at":"1970-01-01T02:00:00.000000000Z","id":4,"slash~1key":"four","slash~1root":false,"value/nested":"d","value/slash~1nested":4.0}
key: {"id":5,"slash~1key":"five","value/nested":"e"}, value: {"canary":"Kringle's","flow_document":"{\"_meta\":{\"uuid\":\"d7a2a680-1de2-11b2-8000-071353030311\"},\"canary\":\"Kringle's\",\"id\":5,\"slash/key\":\"five\",\"slash/root\":true,\"value\":{\"nested\":\"e\",\"slash/nested\":5.0}}","flow_published_at":"1970-01-01T02:00:01.000000000Z","id":5,"slash~1key":"five","slash~1root":true,"value/nested":"e","value/slash~1nested":5.0}
key: {"id":6,"slash~1key":"six","value/nested":"f"}, value: {"canary":"grosbeak's","flow_document":"{\"_meta\":{\"uuid\":\"d83b3d00-1de2-11b2-8000-071353030311\"},\"canary\":\"grosbeak's\",\"id\":6,\"slash/key\":\"six\",\"slash/root\":false,\"value\":{\"nested\":\"f\",\"slash/nested\":6.0}}","flow_published_at":"1970-01-01T02:00:02.000000000Z","id":6,"slash~1key":"six","slash~1root":false,"value/nested":"f","value/slash~1nested":6.0}

************
Topic: test_topic_json_3
************
key: {"id":1,"slash~1key":"one","value/nested":"a"}, value: {"canary":"amputation's","flow_document":"{\"_meta\":{\"uuid\":\"7545a800-1dda-11b2-8000-071353030311\"},\"canary\":\"amputation's\",\"id\":1,\"slash/key\":\"one\",\"slash/root\":true,\"value\":{\"nested\":\"a\",\"slash/nested\":1.0}}","flow_published_at":"1970-01-01T01:00:00.000000000Z","id":1,"slash~1key":"one","slash~1root":true,"value/nested":"a","value/slash~1nested":1.0}
key: {"id":2,"slash~1key":"two","value/nested":"b"}, value: {"canary":"armament's","flow_document":"{\"_meta\":{\"uuid\":\"75de3e80-1dda-11b2-8000-071353030311\"},\"canary\":\"armament's\",\"id\":2,\"slash/key\":\"two\",\"slash/root\":false,\"value\":{\"nested\":\"b\",\"slash/nested\":2.0}}","flow_published_at":"1970-01-01T01:00:01.000000000Z","id":2,"slash~1key":"two","slash~1root":false,"value/nested":"b","value/slash~1nested":2.0}
key: {"id":3,"slash~1key":"three","value/nested":"c"}, value: {"canary":"splatters","flow_document":"{\"_meta\":{\"uuid\":\"7676d500-1dda-11b2-8000-071353030311\"},\"canary\":\"splatters\",\"id\":3,\"slash/key\":\"three\",\"slash/root\":true,\"value\":{\"nested\":\"c\",\"slash/nested\":3.0}}","flow_published_at":"1970-01-01T01:00:02.000000000Z","id":3,"slash~1key":"three","slash~1root":true,"value/nested":"c","value/slash~1nested":3.0}
key: {"id":4,"slash~1key":"four","value/nested":"d"}, value: {"canary":"strengthen","flow_document":"{\"_meta\":{\"uuid\":\"d70a1000-1de2-11b2-8000-071353030311\"},\"canary\":\"strengthen\",\"id\":4,\"slash/key\":\"four\",\"slash/root\":false,\"value\":{\"nested\":\"d\",\"slash/nested\":4.0}}","flow_published_at":"1970-01-01T02:00:00.000000000Z","id":4,"slash~1key":"four","slash~1root":false,"value/nested":"d","value/slash~1nested":4.0}
key: {"id":5,"slash~1key":"five","value/nested":"e"}, value: {"canary":"Kringle's","flow_document":"{\"_meta\":{\"uuid\":\"d7a2a680-1de2-11b2-8000-071353030311\"},\"canary\":\"Kringle's\",\"id\":5,\"slash/key\":\"five\",\"slash/root\":true,\"value\":{\"nested\":\"e\",\"slash/nested\":5.0}}","flow_published_at":"1970-01-01T02:00:01.000000000Z","id":5,"slash~1key":"five","slash~1root":true,"value/nested":"e","value/slash~1nested":5.0}
key: {"id":6,"slash~1key":"six","value/nested":"f"}, value: {"canary":"grosbeak's","flow_document":"{\"_meta\":{\"uuid\":\"d83b3d00-1de2-11b2-8000-071353030311\"},\"canary\":\"grosbeak's\",\"id\":6,\"slash/key\":\"six\",\"slash/root\":false,\"value\":{\"nested\":\"f\",\"slash/nested\":6.0}}","flow_published_at":"1970-01-01T02:00:02.000000000Z","id":6,"slash~1key":"six","slash~1root":false,"value/nested":"f","value/slash~1nested":6.0}
